File: packages/server-sdk/src/webhooks.ts
/**
 * Webhook delivery verification.
 *
 * Outbound deliveries are signed with HMAC-SHA256 over the **raw request
 * body** and sent as `X-PromptHash-Signature: sha256=<hex>`. The signature
 * covers the whole JSON envelope, including `timestamp` and `deliveryId`, so
 * verifying the signature is enough to trust those fields.
 *
 * Always verify against the raw body — never a re-serialised object, because
 * whitespace and key order change the digest.
 *
 * Mirrors `signWebhookPayload` / `verifyWebhookSignature` in
 * `server/src/services/webhookDispatcher.ts`.
 */

import { createHmac, timingSafeEqual } from "node:crypto";

export const SIGNATURE_HEADER = "X-PromptHash-Signature";
export const DELIVERY_HEADER = "X-PromptHash-Delivery";
export const EVENT_HEADER = "X-PromptHash-Event";
export const TIMESTAMP_HEADER = "X-PromptHash-Timestamp";
export const SCHEMA_VERSION_HEADER = "X-PromptHash-Schema-Version";
export const SIGNATURE_PREFIX = "sha256=";

/** Default replay window in seconds. */
export const DEFAULT_TOLERANCE_SECONDS = 300;

/** Shape of the outbound webhook body. */
export interface WebhookEnvelope {
  /** Numeric envelope version. */
  version?: number;
  /** Stable date-string identifying the payload schema. */
  schemaVersion?: string;
  event: string;
  deliveryId: string;
  /** ISO-8601 UTC timestamp of dispatch. */
  timestamp: string;
  data: Record<string, unknown>;
}

export type WebhookFailureReason =
  | "missing_secret"
  | "missing_signature"
  | "missing_timestamp"
  | "malformed_signature"
  | "signature_mismatch"
  | "stale_timestamp"
  | "future_timestamp"
  | "invalid_payload"
  | "duplicate_delivery";

/** Raised by {@link verifyWebhook}; `reason` is safe to log and branch on. */
export class WebhookVerificationError extends Error {
  readonly reason: WebhookFailureReason;

  constructor(reason: WebhookFailureReason, message: string) {
    super(message);
    this.name = "WebhookVerificationError";
    this.reason = reason;
  }
}

/** Case-insensitive header bag — Node's `req.headers` works as-is. */
export type HeaderBag =
  | Record<string, string | string[] | undefined>
  | { get(name: string): string | null };

/** Compute the `sha256=<hex>` signature for a raw body. */
export function signWebhookBody(secret: string, body: string): string {
  if (!secret) throw new WebhookVerificationError("missing_secret", "Webhook secret is required.");
  return SIGNATURE_PREFIX + createHmac("sha256", secret).update(body).digest("hex");
}

/**
 * Constant-time signature check. Returns `false` (never throws) for anything
 * that is not exactly the expected digest.
 */
export function verifyWebhookSignature(
  secret: string,
  body: string,
  signature: string | null | undefined,
): boolean {
  if (!secret || !signature) return false;

  const expected = Buffer.from(signWebhookBody(secret, body), "utf8");
  const received = Buffer.from(signature.trim(), "utf8");
  if (expected.length !== received.length) return false;
  return timingSafeEqual(expected, received);
}

export interface VerifyWebhookOptions {
  /** Maximum accepted clock skew in seconds. Defaults to 300. */
  toleranceSeconds?: number;
  /** Injected clock for deterministic tests (epoch milliseconds). */
  now?: number;
  /** When supplied, repeated `deliveryId`s are rejected. */
  replayGuard?: WebhookReplayGuard;
}

/**
 * Verify a delivery and return its parsed envelope.
 *
 * @param secret  The subscription secret returned at registration time
 * @param body    The raw request body exactly as received
 * @param headers The request headers (case-insensitive)
 * @throws {WebhookVerificationError} when anything fails to check out
 */
export function verifyWebhook(
  secret: string,
  body: string,
  headers: HeaderBag,
  options: VerifyWebhookOptions = {},
): WebhookEnvelope {
  if (!secret) {
    throw new WebhookVerificationError("missing_secret", "Webhook secret is required.");
  }

  const signature = readHeader(headers, SIGNATURE_HEADER);
  if (!signature) {
    throw new WebhookVerificationError(
      "missing_signature",
      `${SIGNATURE_HEADER} header is missing.`,
    );
  }
  if (!signature.startsWith(SIGNATURE_PREFIX)) {
    throw new WebhookVerificationError(
      "malformed_signature",
      `${SIGNATURE_HEADER} must start with ${SIGNATURE_PREFIX}.`,
    );
  }
  if (!verifyWebhookSignature(secret, body, signature)) {
    throw new WebhookVerificationError(
      "signature_mismatch",
      "Webhook signature does not match the request body.",
    );
  }

  let envelope: WebhookEnvelope;
  try {
    envelope = JSON.parse(body) as WebhookEnvelope;
  } catch {
    throw new WebhookVerificationError(
      "invalid_payload",
      "Webhook body is not valid JSON.",
    );
  }
  if (!envelope || typeof envelope !== "object") {
    throw new WebhookVerificationError(
      "invalid_payload",
      "Webhook body is not an object.",
    );
  }
  if (typeof envelope.event !== "string" || !envelope.event) {
    throw new WebhookVerificationError(
      "invalid_payload",
      "Webhook envelope is missing `event`.",
    );
  }
  if (typeof envelope.deliveryId !== "string" || !envelope.deliveryId) {
    throw new WebhookVerificationError(
      "invalid_payload",
      "Webhook envelope is missing `deliveryId`.",
    );
  }

  assertFreshness(
    readHeader(headers, TIMESTAMP_HEADER) ?? envelope.timestamp,
    options.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS,
    options.now ?? Date.now(),
  );

  if (options.replayGuard && !options.replayGuard.accept(envelope.deliveryId)) {
    throw new WebhookVerificationError(
      "duplicate_delivery",
      `Delivery ${envelope.deliveryId} has already been processed.`,
    );
  }

  return envelope;
}

/**
 * Bounded set of recently accepted delivery ids, so a replayed (but
 * correctly signed) delivery is rejected. Entries expire with the replay
 * window, and the oldest are evicted once `maxEntries` is reached.
 */
export class WebhookReplayGuard {
  private readonly ttlMs: number;
  private readonly maxEntries: number;
  private readonly seen = new Map<string, number>();

  constructor(options: { toleranceSeconds?: number; maxEntries?: number } = {}) {
    this.ttlMs = (options.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS) * 1000;
    this.maxEntries = Math.max(1, options.maxEntries ?? 10_000);
  }

  /** Records `deliveryId`; returns `false` when it was seen inside the window. */
  accept(deliveryId: string, now: number = Date.now()): boolean {
    this.prune(now);
    if (this.seen.has(deliveryId)) return false;
    this.seen.set(deliveryId, now);
    while (this.seen.size > this.maxEntries) {
      const oldest = this.seen.keys().next();
      if (oldest.done) break;
      this.seen.delete(oldest.value);
    }
    return true;
  }

  /** Number of ids currently retained. */
  get size(): number {
    return this.seen.size;
  }

  private prune(now: number): void {
    for (const [id, at] of this.seen) {
      if (now - at > this.ttlMs) this.seen.delete(id);
    }
  }
}

function assertFreshness(
  raw: string | undefined,
  toleranceSeconds: number,
  now: number,
): void {
  if (!raw) {
    throw new WebhookVerificationError(
      "missing_timestamp",
      "Webhook timestamp is missing.",
    );
  }

  const numeric = /^\d+$/.test(raw);
  const at = numeric ? Number(raw) : Date.parse(raw);
  if (Number.isNaN(at)) {
    throw new WebhookVerificationError(
      "invalid_payload",
      `Webhook timestamp is not parseable: ${raw}`,
    );
  }

  // Accept either epoch milliseconds or epoch seconds.
  const epochMs = numeric && raw.length <= 10 ? at * 1000 : at;
  const skewMs = now - epochMs;
  const toleranceMs = toleranceSeconds * 1000;

  if (skewMs > toleranceMs) {
    throw new WebhookVerificationError(
      "stale_timestamp",
      "Webhook timestamp is outside the accepted replay window.",
    );
  }
  if (-skewMs > toleranceMs) {
    throw new WebhookVerificationError(
      "future_timestamp",
      "Webhook timestamp is too far in the future.",
    );
  }
}

function readHeader(headers: HeaderBag, name: string): string | undefined {
  if (typeof (headers as { get?: unknown }).get === "function") {
    const value = (headers as { get(n: string): string | null }).get(name);
    return value ?? undefined;
  }
  const bag = headers as Record<string, string | string[] | undefined>;
  const direct = bag[name] ?? bag[name.toLowerCase()];
  const value = direct ?? Object.entries(bag).find(
    ([key]) => key.toLowerCase() === name.toLowerCase(),
  )?.[1];
  if (Array.isArray(value)) return value[0];
  return value;
}

File: packages/server-sdk-python/src/prompthash_server_sdk/webhooks.py
"""Webhook delivery verification.

Deliveries are signed with HMAC-SHA256 over the **raw request body** and sent
as ``X-PromptHash-Signature: sha256=<hex>``. The signature covers the whole
JSON envelope, including ``timestamp`` and ``deliveryId``, so verifying it is
enough to trust those fields.

Always verify against the raw body — never a re-serialised object, because
whitespace and key order change the digest.

Mirrors ``signWebhookPayload`` / ``verifyWebhookSignature`` in
``server/src/services/webhookDispatcher.ts``.
"""

from __future__ import annotations

import hashlib
import hmac
import json
from collections.abc import Mapping
from datetime import datetime, timezone
from typing import Any, Optional

SIGNATURE_HEADER = "X-PromptHash-Signature"
DELIVERY_HEADER = "X-PromptHash-Delivery"
EVENT_HEADER = "X-PromptHash-Event"
TIMESTAMP_HEADER = "X-PromptHash-Timestamp"
SCHEMA_VERSION_HEADER = "X-PromptHash-Schema-Version"
SIGNATURE_PREFIX = "sha256="

#: Default replay window in seconds.
DEFAULT_TOLERANCE_SECONDS = 300


class WebhookVerificationError(Exception):
    """Raised by :func:`verify_webhook`; ``reason`` is safe to log and branch on."""

    def __init__(self, reason: str, message: str) -> None:
        super().__init__(message)
        self.reason = reason


def sign_webhook_body(secret: str, body: str) -> str:
    """Compute the ``sha256=<hex>`` signature for a raw body."""
    if not secret:
        raise WebhookVerificationError("missing_secret", "Webhook secret is required.")
    digest = hmac.new(secret.encode("utf-8"), body.encode("utf-8"), hashlib.sha256)
    return SIGNATURE_PREFIX + digest.hexdigest()


def verify_webhook_signature(
    secret: str, body: str, signature: Optional[str]
) -> bool:
    """Constant-time signature check. Returns ``False`` instead of raising."""
    if not secret or not signature:
        return False
    return hmac.compare_digest(sign_webhook_body(secret, body), signature.strip())


def verify_webhook(
    secret: str,
    body: str,
    headers: Any,
    *,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
    now: Optional[float] = None,
    replay_guard: Optional["WebhookReplayGuard"] = None,
) -> dict:
    """Verify a delivery and return its parsed envelope.

    :raises WebhookVerificationError: when anything fails to check out
    """
    if not secret:
        raise WebhookVerificationError("missing_secret", "Webhook secret is required.")

    signature = _read_header(headers, SIGNATURE_HEADER)
    if not signature:
        raise WebhookVerificationError(
            "missing_signature", f"{SIGNATURE_HEADER} header is missing."
        )
    if not signature.startswith(SIGNATURE_PREFIX):
        raise WebhookVerificationError(
            "malformed_signature",
            f"{SIGNATURE_HEADER} must start with {SIGNATURE_PREFIX}.",
        )
    if not verify_webhook_signature(secret, body, signature):
        raise WebhookVerificationError(
            "signature_mismatch",
            "Webhook signature does not match the request body.",
        )

    try:
        envelope = json.loads(body)
    except ValueError as exc:
        raise WebhookVerificationError(
            "invalid_payload", "Webhook body is not valid JSON."
        ) from exc
    if not isinstance(envelope, dict):
        raise WebhookVerificationError("invalid_payload", "Webhook body is not an object.")
    if not isinstance(envelope.get("event"), str) or not envelope["event"]:
        raise WebhookVerificationError(
            "invalid_payload", "Webhook envelope is missing `event`."
        )
    if not isinstance(envelope.get("deliveryId"), str) or not envelope["deliveryId"]:
        raise WebhookVerificationError(
            "invalid_payload", "Webhook envelope is missing `deliveryId`."
        )

    timestamp = _read_header(headers, TIMESTAMP_HEADER) or envelope.get("timestamp")
    _assert_freshness(
        timestamp,
        tolerance_seconds,
        now if now is not None else _now_ms(),
    )

    if replay_guard is not None and not replay_guard.accept(envelope["deliveryId"]):
        raise WebhookVerificationError(
            "duplicate_delivery",
            f"Delivery {envelope['deliveryId']} has already been processed.",
        )

    return envelope


class WebhookReplayGuard:
    """Bounded set of recently accepted delivery ids.

    Entries expire with the replay window and the oldest are evicted once
    ``max_entries`` is reached.
    """

    def __init__(
        self,
        *,
        tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
        max_entries: int = 10_000,
    ) -> None:
        self._ttl_ms = tolerance_seconds * 1000.0
        self._max_entries = max(1, int(max_entries))
        self._seen: dict[str, float] = {}

    def accept(self, delivery_id: str, now: Optional[float] = None) -> bool:
        """Record ``delivery_id``; ``False`` when it was seen inside the window."""
        current = now if now is not None else _now_ms()
        self._prune(current)
        if delivery_id in self._seen:
            return False
        self._seen[delivery_id] = current
        while len(self._seen) > self._max_entries:
            self._seen.pop(next(iter(self._seen)))
        return True

    @property
    def size(self) -> int:
        return len(self._seen)

    def _prune(self, now: float) -> None:
        expired = [key for key, at in self._seen.items() if now - at > self._ttl_ms]
        for key in expired:
            del self._seen[key]


def _now_ms() -> float:
    import time

    return time.time() * 1000.0


def _read_header(headers: Any, name: str) -> Optional[str]:
    if headers is None:
        return None

    getter = getattr(headers, "get", None)
    if callable(getter):
        value = getter(name)
        if isinstance(value, str):
            return value

    if isinstance(headers, Mapping):
        lowered = name.lower()
        for key, value in headers.items():
            if str(key).lower() == lowered and isinstance(value, str):
                return value
    return None


def _parse_timestamp(raw: Any) -> Optional[float]:
    if isinstance(raw, bool) or raw is None:
        return None
    if isinstance(raw, (int, float)):
        value = float(raw)
        return value * 1000.0 if value < 10_000_000_000 else value

    if not isinstance(raw, str):
        return None
    text = raw.strip()
    if not text:
        return None
    if text.isdigit():
        value = float(text)
        # Accept epoch seconds or epoch milliseconds.
        return value * 1000.0 if len(text) <= 10 else value

    iso = text[:-1] + "+00:00" if text.endswith(("Z", "z")) else text
    try:
        parsed = datetime.fromisoformat(iso)
    except ValueError:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.timestamp() * 1000.0


def _assert_freshness(
    raw: Any, tolerance_seconds: int, now_ms: float
) -> None:
    if raw is None or (isinstance(raw, str) and not raw.strip()):
        raise WebhookVerificationError(
            "missing_timestamp", "Webhook timestamp is missing."
        )

    epoch_ms = _parse_timestamp(raw)
    if epoch_ms is None:
        raise WebhookVerificationError(
            "invalid_payload", f"Webhook timestamp is not parseable: {raw}"
        )

    skew_ms = now_ms - epoch_ms
    tolerance_ms = tolerance_seconds * 1000.0
    if skew_ms > tolerance_ms:
        raise WebhookVerificationError(
            "stale_timestamp",
            "Webhook timestamp is outside the accepted replay window.",
        )
    if -skew_ms > tolerance_ms:
        raise WebhookVerificationError(
            "future_timestamp", "Webhook timestamp is too far in the future."
        )
